src/ReactiveDomain.Users/Domain/Subject.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using ReactiveDomain.Messaging;
+using ReactiveDomain.Users.Messages;
+using ReactiveDomain.Util;
+
+namespace ReactiveDomain.Users.Domain
+{
+    /// <summary>
+    /// User authentication and login.
+    /// </summary>
+    public class Subject : AggregateRoot
+    {
+        private Subject()
+        {
+            RegisterEvents();
+        }
+
+        private void RegisterEvents()
+        {
+            Register<SubjectMsgs.SubjectCreated>(@event => Id = @event.SubjectId);
+        }
+
+        /// <summary>
+        /// Create a new subject for identity server.
+        /// </summary>
+        public Subject(
+            Guid id,
+            Guid userId,
+            string subClaim,
+            string authProvider,
+            string authDomain,
+            ICorrelatedMessage source)
+            : this()
+        {
+            Ensure.NotEmptyGuid(id, nameof(id));
+            Ensure.NotNullOrEmpty(authProvider, nameof(authProvider));
+            Ensure.NotNullOrEmpty(authDomain, nameof(authDomain));
+            Ensure.NotNullOrEmpty(subClaim, nameof(subClaim));
+            Ensure.NotNull(source, nameof(source));
+            Ensure.NotEmptyGuid(userId, nameof(userId));
+            Ensure.NotEmptyGuid(source.CorrelationId, nameof(source.CorrelationId));
+            if (source.CausationId == Guid.Empty)
+                Ensure.NotEmptyGuid(source.MsgId, nameof(source.MsgId));
+
+            ((ICorrelatedEventSource)this).Source = source;
```

Comment:
This works, but it does seem cleaner the way we do it in the other aggregates here where we call `: base(source)` and then call `RegisterEvents()` from within this ctor here instead of calling `: this()` and then manually setting the source.

Reply:
they all seem to work the same in this project, not clear what your suggesting

Reply:
Yeah, they all work the same, just thinking it would be good to standardize on one as a way of modeling what others should do when implementing their own aggregates.

This way of manually setting the `Source` was the only way to do it until we added the overload ctor that accepts a correlation source. The new way seems cleaner and seems like it's how we should recommend doing it going forward.

Reply:
The SecuredApplication class does this for its ctor:
```
        public SecuredApplication(
            Guid id,
            Guid defaultPolicyId,
            string defaultClientId,
            string version,
            bool oneRolePerUser,
            ICorrelatedMessage source)
            : base(source)
        {
            Ensure.NotEmptyGuid(id, nameof(id));
            Ensure.NotEmptyGuid(defaultPolicyId, nameof(defaultPolicyId));
            Ensure.NotNullOrEmpty(defaultClientId, nameof(defaultClientId));
            Ensure.NotNullOrEmpty(version, nameof(version));
            Ensure.NotEmptyGuid(source.CorrelationId, nameof(source));
            RegisterEvents();
            Raise(new ApplicationCreated(
                         id,
                         defaultClientId,
                         version,
                         oneRolePerUser));
            Raise(new PolicyCreated(defaultPolicyId, defaultClientId, id, oneRolePerUser));
        }
```
vs what `Subject` does:
```
        public Subject(
            Guid id,
            Guid userId,
            string subClaim,
            string authProvider,
            string authDomain,
            ICorrelatedMessage source)
            : this()
        {
            Ensure.NotEmptyGuid(id, nameof(id));
            Ensure.NotNullOrEmpty(authProvider, nameof(authProvider));
            Ensure.NotNullOrEmpty(authDomain, nameof(authDomain));
            Ensure.NotNullOrEmpty(subClaim, nameof(subClaim));
            Ensure.NotNull(source, nameof(source));
            Ensure.NotEmptyGuid(userId, nameof(userId));
            Ensure.NotEmptyGuid(source.CorrelationId, nameof(source.CorrelationId));
            if (source.CausationId == Guid.Empty)
                Ensure.NotEmptyGuid(source.MsgId, nameof(source.MsgId));

            ((ICorrelatedEventSource)this).Source = source;
            Raise(new SubjectMsgs.SubjectCreated(
                         id,
                         userId,
                         subClaim,
                         authProvider,
                         authDomain));
        }
```
The first one seems cleaner since it means we're not doing that extra validation of `source` or its IDs or the extra cast on `this` in order to set the `Source` property.

In `User.cs` we actually forgot to validate the `CorrelationId` and `CausationId`, which really points to the fragility of that approach.